### ISSUE TITLE
Force Unmount?

### ISSUE THREAD

**Issue by brockoffdev** [hashicorp/vault#760]
Not sure what we want to do about this, but putting it up for discussion (luckily I had a backup of my vault, so was able to revert).

Basically, due to a missing SQL wrapper command in Postgres 8.0.2, I was not able to properly unmount a mount point I had created which was being used to authenticate into Redshift on AWS.  I was able to create a functioning role to CREATE users, but due to Postgresql's backend revocation SQL command that it sends (namely a DROP ROLE parameter, rather than a DROP USER parameter), Vault would not allow me to revoke any credentials; nor would it allow me to unmount the endpoint at all.

While I plan to make a fork of the Postgresql backend that will work with Redshift, this calls into question whether we should be able to force unmount an endpoint via a Root User.  Say, for example, I lose connectivity to that Redshift as we have disposed of it, or something similar with Postgres, then I will have a "dead" endpoint, as it were, that will forever exist (if I didn't have a backup to revert to), as Vault will have been unable to drop the credentials it had created.  We could even have the CLI warn and alert the user what will occur.

Something as simple as
```
vault force-unmount postgresql

==> WARNING: This will unmount the postgresql backend, and leave unrevoked any and all remaining active
credentials in systems which it maintains. Please retype the mount point you wish to delete to verify:
postgresql

Successfully unmounted 'postgresql' 
```

**Comment by jefferai** [comment]
Hi,

There is already a proposed forked backend (#690), although my question would be the same: why can't normal postgres vs redshift be a config parameter that changes only the necessary behavior? 

I think there are two separate ways that this should be tackled and a complete solution would consist of both:

1. Identify the differences between Postgres and Redshift (it can't be that many) so that default revocation and other behavior can be adjusted by setting a config parameter.
2. Add the ability to specify revocation SQL in addition to creation SQL (but with sensible defaults, controlled by the config parameter).

**Comment by brockoffdev** [comment]
Hey @jefferai...

So first thing:  I think the easy fix would be to have the Postgres endpoint simply have an additional API endpoint requirement that would allow you to configure the SQL to be run during the creation of credentials, AND the SQL to be run during revocation.  This would honestly be preferable, either way, as some users may wish to have multiple things undone following a created credential...

So, in my mind, actually having 2 SQL statements being required on initialization of a role would be best.

***Example (where Redshift is a Postgresql mount type):***

```
vault write redshift/roles/readonly \
    sql_create="CREATE USER \"{{name}}\" WITH PASSWORD '{{password}}' VALID UNTIL '{{expiration}}';
    GRANT SELECT ON ALL TABLES IN SCHEMA public TO \"{{name}}\";"
    sql_revoke="REMOVE USER \"{{name}}\";"
Success! Data written to: readshift/roles/readonly
```
This would be an easy fix to the Redshift issue...the "sql_revoke" parameter can even default to the current "REMOVE ROLE" if nothing is submitted.

Now, second...this still leaves us with the problem...what happens when an endpoint becomes corrupted for some reason.  I still think it's worthwhile having a conversation about there being a way to force unmount a mount point.  Otherwise, dead mounts could pile up for those who do not necessarily have a way of returning to a previous Vault state.

Thanks!

**Comment by jefferai** [comment]
I'm deferring the first point above to #690, as that seems like the right place to continue that conversation.

Can you describe what the issues were that you had with unmounting? It sounds like it was simply that there were outstanding leases, but those should be clearable with `revoke-prefix`.

**Comment by jefferai** [comment]
This was fixed in #1135.
